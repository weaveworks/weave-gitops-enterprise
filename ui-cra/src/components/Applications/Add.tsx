import React, { FC } from 'react';
import { PageTemplate } from '../Layout/PageTemplate';
import { SectionHeader } from '../Layout/SectionHeader';
import { ContentWrapper } from '../Layout/ContentWrapper';
import { ApplicationAdd, theme as weaveTheme } from '@weaveworks/weave-gitops';
import { useApplicationsCount } from './utils';
import styled from 'styled-components';

const ApplicationAddWrapper = styled(ApplicationAdd)`
  & .auth-message {
    margin-bottom: ${weaveTheme.spacing.xs};
  }
`;

const WGApplicationDetail: FC = () => {
  const applicationsCount = useApplicationsCount();

  return (
    <PageTemplate documentTitle="WeGO Â· Application Detail">
      <SectionHeader
        path={[
          {
            label: 'Applications',
            url: '/applications',
            count: applicationsCount,
          },
          { label: 'Add' },
        ]}
      />
      <ContentWrapper type="WG">
        <ApplicationAddWrapper />
      </ContentWrapper>
    </PageTemplate>
  );
};

export default WGApplicationDetail;
