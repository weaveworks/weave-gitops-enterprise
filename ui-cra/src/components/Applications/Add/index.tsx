import React, { useCallback, useEffect, useMemo, useState } from 'react';
import styled from 'styled-components';
import { ThemeProvider } from '@material-ui/core/styles';
import { localEEMuiTheme } from '../../../muiTheme';
import { PageTemplate } from '../../Layout/PageTemplate';
import { createDeploymentObjects, renderKustomization } from '../utils';
import { Grid } from '@material-ui/core';
import { ContentWrapper } from '../../Layout/ContentWrapper';
import {
  Button,
  Flex,
  GitRepository,
  Link,
  LoadingPage,
  useListSources,
} from '@weaveworks/weave-gitops';
import { useHistory } from 'react-router-dom';
import { removeToken } from '../../../utils/request';
import useNotifications from '../../../contexts/Notifications';
import { PageRoute } from '@weaveworks/weave-gitops/ui/lib/types';
import AppFields from './form/Partials/AppFields';
import {
  ClusterAutomation,
  RepositoryRef,
} from '../../../cluster-services/cluster_services.pb';
import _ from 'lodash';
import useProfiles from '../../../hooks/profiles';
import { useCallbackState } from '../../../utils/callback-state';
import {
  AppPRPreview,
  ProfilesIndex,
  ClusterPRPreview,
} from '../../../types/custom';
import { validateFormData } from '../../../utils/form';
import { getGitRepoHTTPSURL } from '../../../utils/formatters';
import { Routes } from '../../../utils/nav';
import Preview from '../../Templates/Form/Partials/Preview';
import Profiles from '../../Templates/Form/Partials/Profiles';
import GitOps from '../../Templates/Form/Partials/GitOps';
import CallbackStateContextProvider from '../../../contexts/GitAuth/CallbackStateContext';
import { clearCallbackState, getProviderToken } from '../../GitAuth/utils';
import {
  useGetInitialGitRepo,
  getRepositoryUrl,
} from '../../Templates/Form/utils';
import { getGitRepos } from '../../Clusters';
import {
  expiredTokenNotification,
  useIsAuthenticated,
} from '../../../hooks/gitprovider';

const FormWrapper = styled.form`
  .preview-cta {
    padding: ${({ theme }) => theme.spacing.small}
      ${({ theme }) => theme.spacing.base};
    button {
      width: 200px;
    }
  }
  .preview-loading {
    padding: ${({ theme }) => theme.spacing.base};
  }
  .create-cta {
    padding: ${({ theme }) => theme.spacing.small};
    button {
      width: 200px;
    }
  }
  .create-loading {
    padding: ${({ theme }) => theme.spacing.base};
  }
`;

const SourceLinkWrapper = styled.div`
  padding-top: ${({ theme }) => theme.spacing.medium};
  overflow-x: auto;
`;

interface FormData {
  repo: GitRepository | null;
  provider: string;
  branchName: string;
  pullRequestTitle: string;
  commitMessage: string;
  pullRequestDescription: string;
  source_name: string;
  source_namespace: string;
  source: string;
  source_type: string;
  source_url: string;
  source_branch: string;
  clusterAutomations: {
    name: string;
    namespace: string;
    target_namespace: string;
    cluster_name: string;
    cluster_namespace: string;
    cluster: string;
    cluster_isControlPlane: boolean;
    createNamespace: boolean;
    path: string;
    source_name: string;
    source_namespace: string;
    source: string;
    source_type: string;
    source_url: string;
    source_branch: string;
  }[];
}

function getInitialData(
  callbackState: { state: { formData: FormData } } | null,
  random: string,
) {
  let defaultFormData = {
    repo: null,
    provider: '',
    branchName: `add-application-branch-${random}`,
    pullRequestTitle: 'Add application',
    commitMessage: 'Add application',
    pullRequestDescription: 'This PR adds a new application',
    clusterAutomations: [
      {
        name: '',
        namespace: '',
        target_namespace: '',
        cluster_name: '',
        cluster_namespace: '',
        cluster: '',
        cluster_isControlPlane: false,
        createNamespace: false,
        path: '',
        source_name: '',
        source_namespace: '',
        source: '',
        source_type: '',
        source_url: '',
        source_branch: '',
      },
    ],
  };

  const initialFormData = {
    ...defaultFormData,
    ...callbackState?.state?.formData,
  };

  return { initialFormData };
}

const AddApplication = ({ clusterName }: { clusterName?: string }) => {
  const [loading, setLoading] = useState<boolean>(false);
  const [showAuthDialog, setShowAuthDialog] = useState(false);
  const { setNotifications } = useNotifications();
  const history = useHistory();
  const authRedirectPage = `/applications/create`;
  const [formError, setFormError] = useState<string>('');

  const optionUrl = (url?: string, branch?: string) => {
    const linkText = branch ? (
      <>
        {url}@<strong>{branch}</strong>
      </>
    ) : (
      url
    );
    if (branch) {
      return (
        <Link href={getGitRepoHTTPSURL(url, branch)} newTab>
          {linkText}
        </Link>
      );
    } else {
      return (
        <Link href={getGitRepoHTTPSURL(url)} newTab>
          {linkText}
        </Link>
      );
    }
  };

  const random = useMemo(() => Math.random().toString(36).substring(7), []);

  const callbackState = useCallbackState();

  const { initialFormData } = getInitialData(callbackState, random);

  const [formData, setFormData] = useState<any>(initialFormData);
  const firstAuto = formData.clusterAutomations[0];
  const helmRepo: RepositoryRef = useMemo(() => {
    return {
      name: firstAuto.source_name,
      namespace: firstAuto.source_namespace,
      cluster: {
        name: firstAuto.cluster_name,
        namespace: firstAuto.cluster_namespace,
      },
    };
  }, [firstAuto]);

  const { profiles, isLoading: profilesIsLoading } = useProfiles(
    firstAuto.source_type === 'HelmRepository',
    undefined,
    undefined,
    helmRepo,
  );
  const [updatedProfiles, setUpdatedProfiles] = useState<ProfilesIndex>({});
  const [openPreview, setOpenPreview] = useState(false);
  const [previewLoading, setPreviewLoading] = useState<boolean>(false);
  const [PRPreview, setPRPreview] = useState<
    ClusterPRPreview | AppPRPreview | null
  >(null);
  const { data } = useListSources();
  const gitRepos = React.useMemo(
    () => getGitRepos(data?.result),
    [data?.result],
  );
  const initialGitRepo = useGetInitialGitRepo(null, gitRepos);

  useEffect(() => {
    setUpdatedProfiles({
      ..._.keyBy(profiles, 'name'),
      ...callbackState?.state?.updatedProfiles,
    });
  }, [callbackState?.state?.updatedProfiles, profiles]);

  useEffect(() => clearCallbackState(), []);

  useEffect(() => {
    setFormData((prevState: any) => ({
      ...prevState,
      pullRequestTitle: `Add application ${(formData.clusterAutomations || [])
        .map((a: any) => a.name)
        .join(', ')}`,
    }));
  }, [formData.clusterAutomations]);

  const getKustomizations = useCallback(() => {
    let clusterAutomations: ClusterAutomation[] = [];
    const selectedProfilesList = _.sortBy(
      Object.values(updatedProfiles),
      'name',
    ).filter(p => p.selected);
    if (formData.source_type === 'HelmRepository') {
      for (let kustomization of formData.clusterAutomations) {
        for (let profile of selectedProfilesList) {
          let values: string = '';
          let version: string = '';
          for (let value of profile.values) {
            if (value.selected === true) {
              version = value.version;
              values = value.yaml;
              clusterAutomations.push({
                cluster: {
                  name: kustomization.cluster_name,
                  namespace: kustomization.cluster_namespace,
                },
                isControlPlane: kustomization.cluster_isControlPlane,
                helmRelease: {
                  metadata: {
                    name: profile.name,
                    namespace: profile.namespace,
                  },
                  spec: {
                    chart: {
                      spec: {
                        chart: profile.name,
                        sourceRef: {
                          name: formData.source_name,
                          namespace: formData.source_namespace,
                        },
                        version,
                      },
                    },
                    values,
                  },
                },
              });
            }
          }
        }
      }
    } else {
      clusterAutomations = formData.clusterAutomations.map(
        (kustomization: any) => {
          return {
            cluster: {
              name: kustomization.cluster_name,
              namespace: kustomization.cluster_namespace,
            },
            isControlPlane: kustomization.cluster_isControlPlane,
            kustomization: {
              metadata: {
                name: kustomization.name,
                namespace: kustomization.namespace,
              },
              spec: {
                path: kustomization.path,
                sourceRef: {
                  name: kustomization.source_name,
                  namespace: kustomization.source_namespace,
                },
                targetNamespace: kustomization.target_namespace,
                createNamespace: kustomization.createNamespace,
              },
            },
          };
        },
      );
    }
    return clusterAutomations;
  }, [
    formData.clusterAutomations,
    formData.source_name,
    formData.source_namespace,
    formData.source_type,
    updatedProfiles,
  ]);

  useEffect(() => {
    if (!formData.repo) {
      setFormData((prevState: any) => ({
        ...prevState,
        repo: initialGitRepo,
      }));
    }
  }, [initialGitRepo, formData.repo]);

  const handlePRPreview = useCallback(() => {
    setPreviewLoading(true);
    return renderKustomization({
      clusterAutomations: getKustomizations(),
    })
      .then(data => {
        setOpenPreview(true);
        setPRPreview(data);
      })
      .catch(err =>
        setNotifications([
          {
            message: { text: err.message },
            severity: 'error',
            display: 'bottom',
          },
        ]),
      )
      .finally(() => setPreviewLoading(false));
  }, [setOpenPreview, getKustomizations, setNotifications]);

  const token = getProviderToken(formData.provider);

  const { isAuthenticated, validateToken } = useIsAuthenticated(
    formData.provider,
    token,
  );

  const handleAddApplication = useCallback(() => {
    const payload = {
      head_branch: formData.branchName,
      title: formData.pullRequestTitle,
      description: formData.pullRequestDescription,
      commit_message: formData.commitMessage,
      clusterAutomations: getKustomizations(),
      repositoryUrl: getRepositoryUrl(formData.repo),
      baseBranch: formData.repo.obj.spec.ref.branch
    };
    setLoading(true);
    return validateToken()
      .then(() =>
        createDeploymentObjects(payload, getProviderToken(formData.provider))
          .then(response => {
            setPRPreview(null);
            history.push(Routes.Applications);
            setNotifications([
              {
                message: {
                  component: (
                    <Link href={response.webUrl} newTab>
                      PR created successfully, please review and merge the pull
                      request to apply the changes to the cluster.
                    </Link>
                  ),
                },
                severity: 'success',
              },
            ]);
          })
          .catch(error =>
            setNotifications([
              {
                message: { text: error.message },
                severity: 'error',
                display: 'bottom',
              },
            ]),
          )
          .finally(() => setLoading(false)),
      )
      .catch(() => {
        removeToken(formData.provider);
        setNotifications([expiredTokenNotification]);
      })
      .finally(() => setLoading(false));
  }, [formData, history, getKustomizations, setNotifications, validateToken]);

  const [submitType, setSubmitType] = useState<string>('');

  return useMemo(() => {
    return (
      <ThemeProvider theme={localEEMuiTheme}>
        <PageTemplate
          documentTitle="Add new application"
          path={[
            {
              label: 'Applications',
              url: Routes.Applications,
            },
            { label: 'Add new application' },
          ]}
        >
          <CallbackStateContextProvider
            callbackState={{
              page: authRedirectPage as PageRoute,
              state: {
                formData,
                updatedProfiles,
              },
            }}
          >
            <ContentWrapper>
              <FormWrapper
                noValidate
                onSubmit={event =>
                  validateFormData(
                    event,
                    submitType === 'PR Preview'
                      ? handlePRPreview
                      : handleAddApplication,
                    setFormError,
                    setSubmitType,
                  )
                }
              >
                <Grid container>
                  <Grid item xs={12} sm={10} md={10} lg={8}>
                    {formData.clusterAutomations.map(
                      (
                        automation: FormData['clusterAutomations'][0],
                        index: number,
                      ) => {
                        return (
                          <AppFields
                            context="app"
                            key={index}
                            index={index}
                            formData={formData}
                            setFormData={setFormData}
                            allowSelectCluster
                            clusterName={clusterName}
                            formError={formError}
                          />
                        );
                      },
                    )}
                    {openPreview && PRPreview ? (
                      <Preview
                        context="app"
                        openPreview={openPreview}
                        setOpenPreview={setOpenPreview}
                        PRPreview={PRPreview}
                        sourceType={formData.source_type}
                      />
                    ) : null}
                  </Grid>
                  <Grid item sm={2} md={2} lg={4}>
                    <SourceLinkWrapper>
                      {optionUrl(formData.source_url, formData.source_branch)}
                    </SourceLinkWrapper>
                  </Grid>
                  {formData.source_type === 'HelmRepository' ? (
                    <Profiles
                      cluster={{
                        name: formData.clusterAutomations[0].cluster_name,
                        namespace:
                          formData.clusterAutomations[0].cluster_namespace,
                      }}
                      // Temp fix to hide layers when using profiles in Add App until we update the BE
                      context="app"
                      isLoading={profilesIsLoading}
                      updatedProfiles={updatedProfiles}
                      setUpdatedProfiles={setUpdatedProfiles}
                      helmRepo={helmRepo}
                    />
                  ) : null}
                  <Grid item xs={12} sm={10} md={10} lg={8}>
                    {previewLoading ? (
                      <LoadingPage className="preview-loading" />
                    ) : (
                      <Flex end className="preview-cta">
                        <Button
                          type="submit"
                          onClick={() => setSubmitType('PR Preview')}
                        >
                          PREVIEW PR
                        </Button>
                      </Flex>
                    )}
                  </Grid>
                  <Grid item xs={12} sm={10} md={10} lg={8}>
                    <GitOps
                      formData={formData}
                      setFormData={setFormData}
                      showAuthDialog={showAuthDialog}
                      setShowAuthDialog={setShowAuthDialog}
                      formError={formError}
                      enableGitRepoSelection={true}
                    />
                    {loading ? (
                      <LoadingPage className="create-loading" />
                    ) : (
                      <Flex end className="create-cta">
                        <Button
                          type="submit"
                          onClick={() => setSubmitType('Create app')}
                          disabled={!isAuthenticated}
                        >
                          CREATE PULL REQUEST
                        </Button>
                      </Flex>
                    )}
                  </Grid>
                </Grid>
              </FormWrapper>
            </ContentWrapper>
          </CallbackStateContextProvider>
        </PageTemplate>
      </ThemeProvider>
    );
  }, [
    authRedirectPage,
    formData,
    handleAddApplication,
    loading,
    profilesIsLoading,
    updatedProfiles,
    setUpdatedProfiles,
    showAuthDialog,
    PRPreview,
    openPreview,
    handlePRPreview,
    previewLoading,
    clusterName,
    helmRepo,
    formError,
    submitType,
    isAuthenticated,
  ]);
};

export default ({ ...rest }) => <AddApplication {...rest} />;
