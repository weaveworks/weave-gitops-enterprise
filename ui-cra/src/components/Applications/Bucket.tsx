import { FC } from 'react';
import {
  BucketDetail,
  Kind,
  useGetObject,
  V2Routes,
} from '@weaveworks/weave-gitops';
import { ContentWrapper } from '../Layout/ContentWrapper';
import { PageTemplate } from '../Layout/PageTemplate';
import { Bucket } from '@weaveworks/weave-gitops/ui/lib/objects';
import { EditButton } from '../Templates/Edit/EditButton';

type Props = {
  name: string;
  namespace: string;
  clusterName: string;
};

const WGApplicationsBucket: FC<Props> = props => {
  const { name, namespace, clusterName } = props;
  const {
    data: bucket,
    isLoading,
    error,
  } = useGetObject<Bucket>(name, namespace, Kind.Bucket, clusterName);

  return (
    <PageTemplate
      documentTitle="Bucket"
      path={[
        {
          label: 'Sources',
          url: V2Routes.Sources,
        },
        {
          label: `${props.name}`,
        },
      ]}
    >
      <ContentWrapper
        loading={isLoading}
        errors={
          error ? [{ clusterName, namespace, message: error?.message }] : []
        }
      >
        <BucketDetail
          bucket={bucket}
          customActions={[<EditButton resource={bucket} />]}
          {...props}
        />
      </ContentWrapper>
    </PageTemplate>
  );
};

export default WGApplicationsBucket;
