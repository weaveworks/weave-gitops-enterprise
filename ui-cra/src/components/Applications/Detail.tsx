import React, { FC } from 'react';
import { PageTemplate } from '../Layout/PageTemplate';
import { SectionHeader } from '../Layout/SectionHeader';
import { WGContentWrapper } from '../Layout/ContentWrapper';
import { ApplicationDetail } from '@weaveworks/weave-gitops';
import { useApplicationsCount } from './utils';

const WGApplicationDetail: FC = () => {
  const applicationsCount = useApplicationsCount();

  const queryParams = new URLSearchParams(window.location.search);
  const name = queryParams.get('name');

  return (
    <PageTemplate documentTitle="WeGO Â· Application Detail">
      <SectionHeader
        path={[
          {
            label: 'Applications',
            url: '/applications',
            count: applicationsCount,
          },
          { label: `${name}` },
        ]}
      />
      <WGContentWrapper>
        <ApplicationDetail name={name || ''} />
      </WGContentWrapper>
    </PageTemplate>
  );
};

export default WGApplicationDetail;
