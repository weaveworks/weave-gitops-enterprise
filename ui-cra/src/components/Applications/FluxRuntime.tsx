import {
  FluxRuntime,
  useListFluxRuntimeObjects,
} from '@weaveworks/weave-gitops';
import { FC } from 'react';
import { ContentWrapper } from '../Layout/ContentWrapper';
import { PageTemplate } from '../Layout/PageTemplate';

const WGApplicationsFluxRuntime: FC = () => {
  const { data, isLoading } = useListFluxRuntimeObjects();

  return (
    <PageTemplate
      documentTitle="WeGO Â· Flux Runtime"
      path={[
        {
          label: 'Flux Runtime',
          url: '/flux_runtime',
          count: data?.deployments?.length,
        },
      ]}
    >
      <ContentWrapper errors={data?.errors} loading={isLoading}>
        <FluxRuntime deployments={data?.deployments} />
      </ContentWrapper>
    </PageTemplate>
  );
};

export default WGApplicationsFluxRuntime;
