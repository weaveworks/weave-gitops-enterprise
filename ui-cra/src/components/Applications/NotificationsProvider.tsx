import { FC } from 'react';
import {
  Kind,
  ProviderDetail,
  useGetObject,
  V2Routes,
} from '@weaveworks/weave-gitops';
import { ContentWrapper } from '../Layout/ContentWrapper';
import { PageTemplate } from '../Layout/PageTemplate';
import { Provider } from '@weaveworks/weave-gitops/ui/lib/objects';

type Props = {
  className?: string;
  name: string;
  namespace: string;
  clusterName: string;
};

const WGNotificationsProvider: FC<Props> = ({
  name,
  namespace,
  clusterName,
}) => {
  const { data, isLoading, error } = useGetObject<Provider>(
    name,
    namespace,
    Kind.Provider,
    clusterName,
  );

  return (
    <PageTemplate
      documentTitle="Notification Providers"
      path={[
        {
          label: 'Notification Providers',
          url: V2Routes.Provider,
        },
      ]}
    >
      <ContentWrapper
        loading={isLoading}
        errors={error ? [{ message: error?.message }] : []}
      >
        <ProviderDetail provider={data} />
      </ContentWrapper>
    </PageTemplate>
  );
};

export default WGNotificationsProvider;
