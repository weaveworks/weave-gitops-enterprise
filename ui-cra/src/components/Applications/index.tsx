import { FC, useMemo } from 'react';
import { PageTemplate } from '../Layout/PageTemplate';
import { ContentWrapper } from '../Layout/ContentWrapper';
import {
  AutomationsTable,
  Button,
  Icon,
  IconType,
  LoadingPage,
  useListAutomations,
  useListSources,
} from '@weaveworks/weave-gitops';
import styled from 'styled-components';
import { useHistory } from 'react-router-dom';
import { Routes } from '../../utils/nav';
import OpenedPullRequest from '../Clusters/OpenedPullRequest';
import { getGitRepos } from '../Clusters';

interface Size {
  size?: 'small';
}
const ActionsWrapper = styled.div<Size>`
  display: flex;
  & > .actionButton.btn {
    margin-right: ${({ theme }) => theme.spacing.small};
  }
`;

const WGApplicationsDashboard: FC = () => {
  const { data: automations, isLoading } = useListAutomations();
  const history = useHistory();
  const { data: sources } = useListSources();

  const gitRepos = useMemo(
    () => getGitRepos(sources?.result),
    [sources?.result],
  );
  const handleAddApplication = () => history.push(Routes.AddApplication);

  return (
    <PageTemplate
      documentTitle="Applications"
      path={[
        {
          label: 'Applications',
        },
      ]}
    >
      <ContentWrapper errors={automations?.errors}>
        <div
          style={{
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'space-between',
          }}
        >
          <ActionsWrapper>
            <Button
              id="add-application"
              className="actionButton btn"
              startIcon={<Icon type={IconType.AddIcon} size="base" />}
              onClick={handleAddApplication}
            >
              ADD AN APPLICATION
            </Button>
            <OpenedPullRequest gitRepos={gitRepos}></OpenedPullRequest>
          </ActionsWrapper>
        </div>
        {isLoading ? (
          <LoadingPage />
        ) : (
          <AutomationsTable automations={automations?.result} />
        )}
      </ContentWrapper>
    </PageTemplate>
  );
};

export default WGApplicationsDashboard;
