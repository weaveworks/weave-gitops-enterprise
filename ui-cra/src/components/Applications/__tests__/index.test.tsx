import Applications from '../';

import { MuiThemeProvider } from '@material-ui/core';
import { act, render, RenderResult, screen } from '@testing-library/react';
import { CoreClientContextProvider, theme } from '@weaveworks/weave-gitops';
import { QueryClient, QueryClientProvider } from 'react-query';
import { MemoryRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import EnterpriseClientProvider from '../../../contexts/EnterpriseClient/Provider';
import NotificationsProvider from '../../../contexts/Notifications/Provider';
import RequestContextProvider from '../../../contexts/Request';
import { muiTheme } from '../../../muiTheme';
import {
  CoreClientMock,
  EnterpriseClientMock,
  withContext,
} from '../../../utils/test-utils';

describe('Applications index test', () => {
  let wrap: (el: JSX.Element) => JSX.Element;
  let api: CoreClientMock;

  beforeEach(() => {
    api = new CoreClientMock();
    wrap = withContext([
      [ThemeProvider, { theme: theme }],
      [MuiThemeProvider, { theme: muiTheme }],
      [
        RequestContextProvider,
        { fetch: () => new Promise(accept => accept(null)) },
      ],
      [QueryClientProvider, { client: new QueryClient() }],
      [
        EnterpriseClientProvider,
        {
          api: new EnterpriseClientMock(),
        },
      ],
      [CoreClientContextProvider, { api }],
      [MemoryRouter],
      [NotificationsProvider],
    ]);
  });
  it('renders table rows', async () => {
    api.ListObjectsReturns = {
      objects: [
        {
          payload: JSON.stringify({
            // maybe?
            apiVersion: 'kustomize.toolkit.fluxcd.io/v1beta2',
            kind: 'Kustomization',
            metadata: {
              namespace: 'my-ns',
              name: 'my-kustomization',
            },
            spec: {
              path: './',
              interval: {},
              sourceRef: {},
            },
            status: {
              conditions: [],
              lastAppliedRevision: '',
              lastAttemptedRevision: '',
              inventory: [],
            },
          }),
          clusterName: 'my-cluster',
        },
      ],
    };

    await act(async () => {
      const c = wrap(<Applications />);
      render(c);
    });

    expect(await screen.findByText('my-kustomization')).toBeTruthy();
  });

  describe('snapshots', () => {
    it('loading', async () => {
      await act(async () => {
        const c = wrap(<Applications />);
        const result = render(c);

        expect(result.container).toMatchSnapshot();
      });
    });
    it('success', async () => {
      let result: RenderResult;
      await act(async () => {
        const c = wrap(<Applications />);
        result = await render(c);
      });

      //   @ts-ignore
      expect(result.container).toMatchSnapshot();
    });
  });
});
