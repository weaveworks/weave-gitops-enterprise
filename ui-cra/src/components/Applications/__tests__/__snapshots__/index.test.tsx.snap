// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Applications index test snapshots loading 1`] = `
.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c8 svg {
  fill: !important;
  height: 24px;
  width: 24px;
}

.c8.downward {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c8.upward {
  -webkit-transform: initial;
  -ms-transform: initial;
  transform: initial;
}

.c12 svg {
  fill: !important;
  height: 16px;
  width: 16px;
}

.c12.downward {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c12.upward {
  -webkit-transform: initial;
  -ms-transform: initial;
  transform: initial;
}

.c11.MuiButton-root {
  height: 32px;
  font-size: 12px;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  line-height: 1;
  border-radius: 2px;
  font-weight: 600;
}

.c11.MuiButton-outlined {
  padding: 8px 12px;
  border-color: #d8d8d8;
}

.c6 {
  height: 40px;
  width: 40px;
}

.c6.MuiIconButton-root {
  background-color: #fff;
}

.c6.MuiIconButton-root:hover {
  background-color: #fff;
  color: #009CCC;
}

.c4 {
  padding-right: 12px;
}

.c0 {
  width: 100%;
  margin: 0 auto;
}

.c9 {
  margin: 0 16px;
  padding: 24px;
  background-color: #fff;
  border-radius: 8px 8px 0 0;
}

.c14 {
  padding: calc(24px - 4px) 24px;
  margin: 0 16px;
  background-color: rgba(255,255,255,0.7);
  color: #737373;
  border-radius: 0 0 8px 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c14 a {
  color: #00b3ec;
}

.c2 {
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 20px;
  height: 32px;
}

.c3 {
  margin-right: 8px;
  white-space: nowrap;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: inherit;
  font-size: 20px;
  height: 80px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  padding: 0 12px 0 24px;
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  z-index: 2;
  background: inherit;
}

.c1 .MuiListItemIcon-root {
  min-width: 30px;
}

.c5 svg {
  fill: #00b3ec;
}

.c5 button {
  background-color: #f5f5f5;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c10 > .actionButton.btn {
  margin-right: 12px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="makeStyles-path-10"
          style="align-items: center;"
        >
          <div
            class="c3"
            role="heading"
          >
            <a
              class="makeStyles-link-13"
              href="/applications"
            >
              Applications
            </a>
          </div>
        </div>
      </div>
      <div
        class="c4 c5"
      >
        <button
          aria-haspopup="true"
          class="MuiButtonBase-root MuiIconButton-root c6"
          tabindex="0"
          title="Account settings"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <div
              class="c7 c8"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                />
              </svg>
            </div>
          </span>
        </button>
      </div>
    </div>
    <div
      style="display: flex; flex-direction: column; width: 100%; height: calc(100vh - 80px); overflow-wrap: normal; overflow-x: scroll;"
    >
      <div
        class="c9"
      >
        <div
          style="display: flex; align-items: center; justify-content: space-between;"
        >
          <div
            class="c10"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined c11 actionButton btn MuiButton-outlinedPrimary MuiButton-disableElevation"
              id="add-application"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                <span
                  class="MuiButton-startIcon MuiButton-iconSizeMedium"
                >
                  <div
                    class="c7 c12"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                      />
                    </svg>
                  </div>
                </span>
                ADD AN APPLICATION
              </span>
            </button>
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined c11 actionButton btn MuiButton-outlinedPrimary MuiButton-disableElevation"
              href="/"
              rel="noopener noreferrer"
              tabindex="0"
              target="_blank"
            >
              <span
                class="MuiButton-label"
              >
                <div
                  class="c7 c12 makeStyles-externalIcon-9"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
                    />
                  </svg>
                </div>
                GO TO OPEN PULL REQUESTS
              </span>
            </a>
          </div>
        </div>
        <div
          class="c13 "
        >
          <div
            class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
            role="progressbar"
            style="width: 40px; height: 40px; color: rgb(0, 179, 236);"
          >
            <svg
              class="MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="c14"
      >
        <div>
          Need help? Raise aÂ 
          <a
            href="https://weavesupport.zendesk.com/"
          >
            support ticket
          </a>
        </div>
        <div
          class=""
        >
          Weave GitOps Enterprise 
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Applications index test snapshots success 1`] = `<div />`;
